includes/specials/SpecialUserrights.php

```diff
@@ -256,36 +256,36 @@ function doSaveUserGroups( $user, $add, $remove, $reason = '' ) {
 		);
 
 		$oldGroups = $user->getGroups();
+		sort( $oldGroups );
 		$newGroups = $oldGroups;
 
-		// remove then add groups
-		if( $remove ) {
-			$newGroups = array_diff( $newGroups, $remove );
-			foreach( $remove as $group ) {
-				if ( !$user->removeGroup( $group ) ) {
-					$newGroups[] = $group;
-				}
-			}
-		}
+		$success = true;
 
+		//First add, then remove, otherwise addition may not be possible due to lack of permissions
 		if( $add ) {
 			$newGroups = array_merge( $newGroups, $add );
-			foreach( $add as $group ) {
-				if ( !$user->addGroup( $group ) ) {
-					if ( ( $key = array_search( $group, $newGroups ) ) !== false ) {
-						unset($newGroups[$key]);
-					}
-				}
-			}
+			$success = $user->addGroup( $add ) && $success;
+		}
+
+		if( $remove ) {
+			$newGroups = array_diff( $newGroups, $remove );
+			$success = $user->removeGroup( $remove ) && $success;
 		}
 
 		$newGroups = array_unique( $newGroups );
+		sort( $newGroups );
+
+		if( !$success ) {
+			$newGroups = $user->getGroups();
+			sort( $newGroups );
+			\Wikia\Logger\WikiaLogger::instance()->error(
+				'Error occurred while changing groups. Groups requested to add: ' . json_encode( $add )
+				. ", groups request to remove: " . json_encode( $remove ) );
+		}
 
-		// Ensure that caches are cleared
-		$user->invalidateCache();
```

Comment:
It's being done inside of UserPermissionsService

https://github.com/Wikia/app/blob/dev/lib/Wikia/src/Service/User/Permissions/PermissionsServiceImpl.php#L249